### ISSUE TITLE
Cannot have type union containing void ( to be used as return type only )

### ISSUE THREAD

**Issue by tonyhallett** [microsoft/dtslint#256]
for instance 

```
type ReturnType = void | Promise<void>;
interface Matchers<R>{
 matcher1(): R,
 matcher2(): R
}
```
Use the `void` type for return types only. Otherwise, use `undefined`. See: https://github.com/Microsoft/dtslint/blob/master/docs/void-return.md

Note that just having `type VoidType = void` without referencing the type anywhere in the source code will throw the error.

**Comment by Maxim-Mazurok** [comment]
That's by design, I think

**Comment by tonyhallett** [comment]
So the return type should be

```type ReturnType = void | Promise<undefined>; //?```

Given that ```Promise<void>``` is the default type inferred by the TypeScript compiler when it sees an async function that does not return a avalue - I would have thought that ```Promise<void>``` should be acceptable.

**Comment by Maxim-Mazurok** [comment]
No, it look like the return type should be `type ReturnType = undefined | Promise<void>;`
